src/messages/mod.rs

```diff
@@ -102,108 +151,132 @@ impl SignedRoutingMessage {
+                let signed_bytes = serialise(&self.content)?;
+                if !security_metadata.verify_sig(&signed_bytes) {
+                    return Err(RoutingError::FailedSignature);
+                }
+                if !security_metadata.validate_proof() {
+                    return Err(RoutingError::InvalidProvingSection);
+                }
+                Ok(())
             }
         }
-        Ok(())
-    }
-
-    /// Returns the previous hop: if that hop can be trusted, the message can be trusted, too.
-    pub fn previous_hop(&self) -> Option<&SectionInfo> {
-        self.proving_sections
-            .last()
-            .map(|ps| &ps.sec_info)
-            .or_else(|| self.src_section.as_ref())
-    }
-
-    /// Removes the last hop section from the message. Returns `true` if a hop was removed, and
-    /// `false` if there are no more hops left other than the sending section itself.
-    pub fn pop_previous_hop(&mut self) -> bool {
-        self.proving_sections.pop().is_some()
     }
 
-    /// Returns the list of section infos from this message's proof chain.
-    pub fn section_infos(&self) -> impl Iterator<Item = &SectionInfo> {
-        self.src_section
-            .iter()
-            .chain(self.proving_sections.iter().map(|ps| &ps.sec_info))
-    }
-
-    /// Returns the chain of proving sections, from the recipient back to the one proving the
-    /// sending section.
-    pub fn proving_sections(&self) -> &Vec<ProvingSection> {
-        &self.proving_sections
+    /// Checks if the message can be trusted according to the Chain
+    pub fn check_trust(&self, chain: &Chain) -> bool {
+        match self.security_metadata {
+            SecurityMetadata::Full(ref security_metadata) => {
+                chain.check_trust(security_metadata.prefix(), security_metadata.proof_chain())
+            }
+            SecurityMetadata::None => true,
+            SecurityMetadata::Partial(_) => false,
+        }
     }
 
     /// Returns the source section that signed the message itself.
     pub fn source_section(&self) -> Option<&SectionInfo> {
         self.src_section.as_ref()
     }
 
-    /// Returns whether the message is signed by the given public ID.
-    #[cfg(test)]
-    pub fn signed_by(&self, pub_id: &PublicId) -> bool {
-        self.signatures.contains_id(pub_id)
-    }
-
     /// Returns the number of nodes in the source authority.
     pub fn src_size(&self) -> usize {
         self.src_section.as_ref().map_or(0, |si| si.members().len())
     }
 
     /// Adds a proof if it is new, without validating it.
-    pub fn add_proof(&mut self, proof: Proof) {
-        if self.content.src.is_multiple() && self.is_sender(proof.pub_id()) {
-            let _ = self.signatures.add_proof(proof);
+    pub fn add_signature_share(
+        &mut self,
+        pk_share: BlsPublicKeyShare,
+        sig_share: BlsSignatureShare,
+    ) {
+        if let SecurityMetadata::Partial(ref mut partial) = self.security_metadata {
```

Comment:
only functional for `partial` ?
also, shall it also have a log_or_panic as the other places in the case of `full` ?

Reply:
For the other, nothing to do.

No log or panic, suggested in original PR, but this function is also called in other cases where it should do nothing (for now at least, we will probably need to clean up this code further as we completing the work).